### PR TITLE
fix(hogql): Optimize querying for date frame in trends actors query

### PR DESCRIPTION
**Pull request by webjunkie**
## Problem

The query runner produces a query with e.g. `toStartOfWeek(timestamp) = {our_time_frame}` which has performance implications because seemingly CH must compute all timestamps and can only then match.

## Changes

- calculate the time range in Python
- because for example  
  
      toStartOfWeek(2024-03-21 12:01:10) = 2024-03-18   
  is the same as saying   
  
      2024-03-21 12:01:10 >= 2024-03-18 00:00:00 and 2024-03-21 12:01:10 < 2024-03-25 00:00:00

## Does this work well for both Cloud and self-hosted?

No impact

## How did you test this code?

- checked modals and tests

### CODE REVIEW

File: posthog/hogql_queries/utils/query_date_range.py [hunk 1]
```diff
@@ -257,6 +261,11 @@ def to_placeholders(self) -> Dict[str, ast.Expr]:
             else self.date_from_as_hogql(),
         }
 
+    def interval_bounds_from_str(self, time_frame: str) -> tuple[datetime, datetime]:
+        date_from = parse(time_frame)
+        date_to = date_from + self.interval_relativedelta()
+        return date_from, date_to
```
Comment (mariusandra) [inline posthog/hogql_queries/utils/query_date_range.py:267]: I wonder if this is susceptible to the same issue that's fixed here: https://github.com/PostHog/posthog/pull/21042

Do we need to worry about timezones or DST anywhere here?

**Review by mariusandra — APPROVED** [review]
👍

File: posthog/hogql_queries/utils/query_date_range.py (hunk 1 above)
Comment (webjunkie) [reply posthog/hogql_queries/utils/query_date_range.py:267]: Good question.... 

If the query like the following, shouldn't it be safe?

`and(greaterOrEquals(toTimeZone(e.timestamp, 'America/Los_Angeles'), toDateTime64('2024-03-18 00:00:00.000000', 6, 'America/Los_Angeles')), less(toTimeZone(e.timestamp, 'America/Los_Angeles'), toDateTime64('2024-03-19 00:00:00.000000', 6, 'America/Los_Angeles')))`

Comment (mariusandra) [reply posthog/hogql_queries/utils/query_date_range.py:267]: Indeed, that should be. My question was if Python gets the calculation correct. Seems so 🤷

Comment (webjunkie) [reply posthog/hogql_queries/utils/query_date_range.py:267]: I just saw I need to push one fix to pass the TZ through, and I can check the relative delta over DST changes as well.

**Pull request merged**